core/shared/src/main/scala/zio/prelude/Semiring.scala

```diff
@@ -0,0 +1,398 @@
+package zio.prelude
+
+import scala.annotation.tailrec
+
+/**
+ * `Semiring` is a data type that represents some notion of "events" that can
+ * take place in parallel or in sequence. For example, a semiring
+ * parameterized on some error type could be used to model the potentially
+ * multiple ways that an application can fail. On the other hand, a semiring
+ * parameterized on some request type could be used to model a collection of
+ * requests to external data sources, some of which could be executed in
+ * parallel and some of which must be executed sequentially.
+ */
+sealed trait Semiring[+Z <: Unit, +A] { self =>
+
+  /**
+   * Combines this collection of events with that collection of events to
+   * return a new collection of events that represents this collection of
+   * events in parallel with that collection of events.
+   */
+  final def &&[Z1 >: Z <: Unit, A1 >: A](that: Semiring[Z1, A1]): Semiring[Z1, A1] =
+    Semiring.Both(self, that)
+
+  /**
+   * A symbolic alias for `zipRight`.
+   */
+  final def *>[Z1 >: Z <: Unit, B](that: Semiring[Z1, B]): Semiring[Z1, B] =
+    zipRight(that)
+
+  /**
+   * Combines this collection of events with that collection of events to
+   * return a new collection of events that represents this collection of
+   * events followed by that collection of events.
+   */
+  final def ++[Z1 >: Z <: Unit, A1 >: A](that: Semiring[Z1, A1]): Semiring[Z1, A1] =
+    Semiring.Then(self, that)
+
+  /**
+   * A symbolic alias for `zipLeft`.
+   */
+  final def <*[Z1 >: Z <: Unit, B](that: Semiring[Z1, B]): Semiring[Z1, A] =
+    zipLeft(that)
+
+  /**
+   * A symbolic alias for `zip`.
+   */
+  final def <*>[Z1 >: Z <: Unit, B](that: Semiring[Z1, B]): Semiring[Z1, (A, B)] =
+    zip(that)
+
+  /**
+   * Maps the events in this collection of events to the specified constant
+   * value.
+   */
+  final def as[B](b: B): Semiring[Z, B] =
+    map(_ => b)
+
+  /**
+   * Returns the first event in this collection of events. If multiple events
+   * occur in parallel and before any other events then any of these events
+   * may be returned.
+   */
+  @tailrec
+  final def first(implicit ev: Z <:< Nothing): A =
+    (self: @unchecked) match {
+      case Semiring.Both(left, _) => left.first
+      case Semiring.Then(left, _) => left.first
+      case Semiring.Single(value) => value
+    }
+
+  /**
+   * Constructs a new collection of events for each event in this collection of
+   * events, collecting them back into a single collection of events.
+   */
+  final def flatMap[Z1 >: Z <: Unit, B](f: A => Semiring[Z1, B]): Semiring[Z1, B] =
+    fold(Semiring.empty, f)(_ ++ _, _ && _).asInstanceOf[Semiring[Z1, B]]
```

Comment:
We know that the resulting collection can only be empty if `Z` or `Z1` is `Unit` but we still have to handle the case where the initial collection is empty so we have to cast here.